File: cpp2nim/analyzer.py
"""
Dependency analysis for cpp2nim.

This module analyzes dependencies between parsed headers and determines
which types should be moved to shared type files.
"""

import os
from dataclasses import dataclass, field
from typing import Any

from .config import Config, get_global_option
from .models import ParseResult, ParsedHeader
from .utils import flatten_namespace


@dataclass
class AnalysisResult:
    """Result of dependency analysis.
    
    Attributes:
        file_relationships: For each file, maps other files to the types they provide.
        shared_types: Types that should be in the shared types file.
        type_renames: Mapping of types that need renaming to avoid conflicts.
        import_graph: For each file, the files it needs to import.
    """
    file_relationships: dict[str, dict[str, set[str]]] = field(default_factory=dict)
    shared_types: set[str] = field(default_factory=set)
    type_renames: dict[str, str] = field(default_factory=dict)
    import_graph: dict[str, set[str]] = field(default_factory=dict)


class DependencyAnalyzer:
    """Analyze dependencies between parsed headers.
    
    This class determines which types are shared between files and need
    to be moved to a common types file.
    
    Example:
        >>> analyzer = DependencyAnalyzer(config)
        >>> result = analyzer.analyze(parse_result)
        >>> print(result.shared_types)
    """
    
    def __init__(self, config: Config | None = None):
        """Initialize the analyzer.
        
        Args:
            config: Configuration settings.
        """
        self.config = config or Config()
    
    def analyze(self, parse_result: ParseResult) -> AnalysisResult:
        """Perform full dependency analysis.
        
        Args:
            parse_result: The result from parsing headers.
            
        Returns:
            AnalysisResult with dependency information.
        """
        result = AnalysisResult()
        
        # Compute file relationships
        result.file_relationships = self._compute_relationships(parse_result)
        
        # Identify shared types
        result.shared_types = self._compute_shared_types(parse_result, result.file_relationships)
        
        # Compute type renames
        result.type_renames = self._compute_renames(parse_result, result.shared_types)
        
        # Build import graph
        result.import_graph = self._compute_imports(parse_result, result.file_relationships)
        
        return result
    
    def _compute_relationships(self, parse_result: ParseResult) -> dict[str, dict[str, set[str]]]:
        """Compute which files provide which types to which other files."""
        relationships: dict[str, dict[str, set[str]]] = {}
        filenames = set(parse_result.headers.keys())
        
        for filename in filenames:
            missing = parse_result.all_missing.get(filename, set())
            file_deps: dict[str, set[str]] = {}
            
            for other_file in filenames:
                if other_file == filename:
                    continue
                
                provides = parse_result.all_provides.get(other_file, set())
                
                # Direct matches
                found = missing.intersection(provides)
                
                # Enum value matches (provides may contain (value, enum) tuples)
                enum_values = [k[0] for k in provides if isinstance(k, tuple)]
                found2 = missing.intersection(enum_values)
                
                enum_found = []
                if found2:
                    for item in found2:
                        for k in provides:
                            if isinstance(k, tuple) and item == k[0]:
                                enum_found.append(k[1])
                enum_found = list(set(enum_found))
                
                if found or enum_found:
                    file_deps[other_file] = set(list(found) + enum_found)
            
            relationships[filename] = file_deps
        
        return relationships
    
    def _compute_shared_types(
        self, 
        parse_result: ParseResult,
        relationships: dict[str, dict[str, set[str]]]
    ) -> set[str]:
        """Determine which types should be in the shared types file."""
        shared: set[str] = set()
        
        # Types needed by multiple files
        for _, deps in relationships.items():
            for _, types in deps.items():
                shared.update(types)
        
        # Add dependencies of shared types
        all_shared = set(shared)
        for type_name in shared:
            deps = self._find_type_dependencies(type_name, parse_result)
            all_shared.update(deps)
        
        # Add forced shared types from config
        all_shared.update(self.config.force_shared_types)
        
        return all_shared
    
    def _find_type_dependencies(
        self, 
        type_name: str, 
        parse_result: ParseResult,
        visited: set[str] | None = None
    ) -> set[str]:
        """Find all types that a given type depends on.
        
        Args:
            type_name: The type to find dependencies for.
            parse_result: Parsed header data.
            visited: Already visited types (for cycle detection).
            
        Returns:
            Set of dependent type names.
        """
        if visited is None:
            visited = set()
        
        if type_name in visited:
            return set()
        visited.add(type_name)
        
        deps: set[str] = set()
        
        # Collect all types from all headers
        all_classes = []
        all_enums = []
        
        for header in parse_result.headers.values():
            for struct in header.structs:
                all_classes.append((struct.fully_qualified, struct.underlying_deps))
            for cls in header.classes:
                all_classes.append((cls.fully_qualified, []))
            for typedef in header.typedefs:
                all_classes.append((typedef.fully_qualified, typedef.underlying_deps))
            for enum in header.enums:
                all_enums.append(enum.fully_qualified)
        
        classes_fully = [c[0] for c in all_classes]
        
        # Find the type and its dependencies
        for fully_qualified, underlying_deps in all_classes:
            if fully_qualified != type_name:
                continue
            
            for dep in underlying_deps:
                if dep in all_enums:
                    deps.add(dep)
                elif dep in classes_fully:
                    if dep not in deps:
                        new_deps = self._find_type_dependencies(dep, parse_result, visited)
                        deps.add(dep)
                        deps.update(new_deps)
        
        return deps
    
    def _compute_renames(
        self, 
        parse_result: ParseResult,
        shared_types: set[str]
    ) -> dict[str, str]:
        """Generate new names for types that would otherwise conflict."""
        renames: dict[str, str] = {}
        
        # Find types in shared that have the same short name
        short_names: dict[str, list[str]] = {}
        
        for type_name in shared_types:
            short = type_name.split("::")[-1]
            if short not in short_names:
                short_names[short] = []
            short_names[short].append(type_name)
        
        # Rename types with conflicting short names
        root_namespace = get_global_option("root_namespace")
        
        for short, full_names in short_names.items():
            if len(full_names) > 1:
                for full in full_names:
                    new_name = self._get_new_name(full, list(renames.values()), root_namespace)
                    renames[full] = new_name
        
        return renames
    
    def _get_new_name(
        self, 
        full_name: str, 
        existing_names: list[str],
        root_namespace: str | None
    ) -> str:
        """Generate a new unique name for a type.
        
        Args:
            full_name: Fully qualified C++ name.
            existing_names: Names already in use.
            root_namespace: Root namespace to strip.
            
        Returns:
            New unique name.
        """
        parts = full_name.split("::")
        if root_namespace and parts and parts[0] == root_namespace:
            return '_'.join(parts[1:])
        return flatten_namespace(full_name)
    
    def _compute_imports(
        self, 
        parse_result: ParseResult,
        relationships: dict[str, dict[str, set[str]]]
    ) -> dict[str, set[str]]:
        """Build the import graph for generated files."""
        imports: dict[str, set[str]] = {}
        
        for filename, deps in relationships.items():
            file_imports: set[str] = set()
            for dep_file in deps.keys():
                # Convert path to module name
                basename = os.path.basename(dep_file)
                module_name = os.path.splitext(basename)[0]
                file_imports.add(module_name)
            imports[filename] = file_imports
        
        return imports


def _relationships(
    data: list[tuple[Any, ...]],
    provides: dict[str, set[str]],
    missing: dict[str, set[str]]
) -> dict[str, dict[str, set[str]]]:
    """Compute file relationships (legacy format).
    
    Args:
        data: Legacy data format from parsing.
        provides: Types provided by each file.
        missing: Missing dependencies for each file.
        
    Returns:
        Relationship mapping.
    """
    relationships: dict[str, dict[str, set[str]]] = {}
    filenames = set(item[0] for item in data)
    
    for filename in filenames:
        file_missing = missing.get(filename, set())
        file_deps: dict[str, set[str]] = {}
        
        for other in filenames:
            if other == filename:
                continue
            
            file_provides = provides.get(other, set())
            found = file_missing.intersection(file_provides)
            
            # Handle enum tuples
            enum_values = [k[0] for k in file_provides if isinstance(k, tuple)]
            found2 = file_missing.intersection(enum_values)
            
            enum_found = []
            if found2:
                for item in found2:
                    for k in file_provides:
                        if isinstance(k, tuple) and item == k[0]:
                            enum_found.append(k[1])
            enum_found = list(set(enum_found))
            
            if found or enum_found:
                file_deps[other] = set(list(found) + enum_found)
        
        relationships[filename] = file_deps
    
    return relationships


def find_dependencies(
    obj: str,
    data: list[tuple[Any, ...]],
    refobj: str | None = None,
    rec: set[str] | None = None
) -> set[str]:
    """Find all dependencies of a type (legacy format).
    
    Args:
        obj: Type name to find dependencies for.
        data: Legacy parsed data.
        refobj: Reference object (unused).
        rec: Recursion tracking set.
        
    Returns:
        Set of dependency type names.
    """
    if rec is None:
        rec = set()
    
    if obj in rec:
        return set()
    rec.add(obj)
    
    deps: set[str] = set()
    
    # Find class/typedef/struct indices
    idx_class = [j for j in range(len(data)) if len(data[j]) > 2 and data[j][2] in ["class", "typedef", "struct"]]
    classes_fully = [data[j][4]["fully_qualified"] if len(data[j]) > 4 else "" for j in idx_class]
    classes = [data[j][3] if len(data[j]) > 3 else "" for j in idx_class]
    
    # Find enum indices
    idx_enum = [j for j in range(len(data)) if len(data[j]) > 2 and data[j][2] == "enum"]
    enums_fully = [data[j][3] if len(data[j]) > 3 else "" for j in idx_enum]
    enums = ["enum " + (data[j][4]["name"] if len(data[j]) > 4 else "") for j in idx_enum]
    
    for i in range(len(data)):
        if len(data[i]) < 3:
            continue
        if data[i][2] not in ["typedef", "class", "struct"]:
            continue
        if len(data[i]) < 5:
            continue
        
        values = data[i][4]
        if values.get("fully_qualified") != obj:
            continue
        
        # Check underlying dependencies
        if "underlying_deps" in values:
            for dep in values["underlying_deps"]:
                if dep in enums_fully:
                    deps.add(dep)
                elif dep in enums:
                    k = enums.index(dep)
                    k = idx_enum[k]
                    full_name = data[k][3] if len(data[k]) > 3 else ""
                    deps.add(full_name)
                elif dep in classes_fully:
                    if dep not in deps:
                        new_deps = find_dependencies(dep, data, obj, rec)
                        deps.add(dep)
                        deps.update(new_deps)
                elif dep in classes:
                    k = classes.index(dep)
                    k = idx_class[k]
                    full_name = data[k][4]["fully_qualified"] if len(data[k]) > 4 else ""
                    if full_name not in deps:
                        new_deps = find_dependencies(full_name, data, obj, rec)
                        deps.add(full_name)
                        deps.update(new_deps)
        
        # Check template params
        if "template_params" in values and values["template_params"]:
            for param in values["template_params"]:
                if isinstance(param, tuple):
                    for k in range(len(enums_fully)):
                        enum_type = enums_fully[k]
                        if enum_type.endswith(param[1]) and "::" in param[1]:
                            if enum_type not in deps:
                                new_deps = find_dependencies(enum_type, data, obj, rec)
                                deps.add(enum_type)
                                deps.update(new_deps)
    
    return deps


def move_to_shared_types(
    new_filename: str,
    data: list[tuple[Any, ...]],
    root: str,
    relations: dict[str, dict[str, set[str]]] | None = None,
    force_shared: list[str] | None = None
) -> list[tuple[Any, ...]]:
    """Move shared types to a common file (legacy format).
    
    Args:
        new_filename: The shared types filename.
        data: Legacy parsed data (modified in place).
        root: Root directory path.
        relations: File relationships.
        force_shared: Types to force into shared file.
        
    Returns:
        Modified data with import statements added.
    """
    relations = relations or {}
    force_shared = force_shared or []
    
    # Collect objects that need to be moved
    objects: set[str] = set()
    for _, deps in relations.items():
        for _, type_set in deps.items():
            objects.update(type_set)
    
    # Add dependencies
    all_objects = set(objects)
    for obj in objects:
        deps = find_dependencies(obj, data, new_filename)
        all_objects.update(deps)
    
    all_objects.update(force_shared)
    
    # Move items to new file
    new_imports: set[str] = set()
    for obj in all_objects:
        for i in range(len(data)):
            if len(data[i]) < 3:
                continue
            
            item_type = data[i][2] if len(data[i]) > 2 else None
            
            if item_type == "const" and len(data[i]) > 3:
                values = data[i][3]
                for item in values.get("items", []):
                    if item.get("name") == obj:
                        new_imports.add(data[i][0])
                        data[i] = (new_filename,) + data[i][1:]
            
            elif item_type == "enum" and len(data[i]) > 3:
                name = data[i][3]
                if name == obj:
                    new_imports.add(data[i][0])
                    data[i] = (new_filename,) + data[i][1:]
            
            elif item_type in ["typedef", "class", "struct"] and len(data[i]) > 4:
                values = data[i][4]
                if values.get("fully_qualified") == obj:
                    new_imports.add(data[i][0])
                    data[i] = (new_filename,) + data[i][1:]
    
    # Add import statements
    new_data = []
    for imp in new_imports:
        new_data.append((imp, None, "import", [new_filename]))
    
    return new_data + data

File: nim/src/cpp2nim/analyzer.nim
## Dependency analysis for cpp2nim.
##
## Analyzes dependencies between parsed headers and determines which types
## should be moved to shared type files.

import std/[sets, tables, options, os, strutils, algorithm, sequtils]
import ./config
import ./models
import ./utils

type
  AnalysisResult* = object
    ## Result of dependency analysis.
    ##
    ## Fields:
    ##   fileRelationships: For each file, maps other files to the types they provide.
    ##   sharedTypes: Types that should be in the shared types file.
    ##   typeRenames: Mapping of types that need renaming to avoid conflicts.
    ##   importGraph: For each file, the files it needs to import.
    ##   baseClasses: Classes that are used as base classes (need of RootObj).
    fileRelationships*: Table[string, Table[string, HashSet[string]]]
    sharedTypes*: HashSet[string]
    typeRenames*: Table[string, string]
    importGraph*: Table[string, HashSet[string]]
    baseClasses*: HashSet[string]

  DependencyAnalyzer* = object
    ## Analyze dependencies between parsed headers.
    ##
    ## Determines which types are shared between files and need to be
    ## moved to a common types file.
    ##
    ## Example:
    ##   let analyzer = initDependencyAnalyzer(config)
    ##   let result = analyzer.analyze(parseResult)
    ##   echo result.sharedTypes
    config*: Config


proc initAnalysisResult*(): AnalysisResult =
  ## Create an empty AnalysisResult.
  AnalysisResult()


proc initDependencyAnalyzer*(config: Config = defaultConfig()): DependencyAnalyzer =
  ## Initialize a dependency analyzer.
  DependencyAnalyzer(config: config)


proc computeRelationships(self: DependencyAnalyzer, parseResult: ParseResult): Table[string, Table[string, HashSet[string]]] =
  ## Compute which files provide which types to which other files.
  var filenames: HashSet[string]
  for k in parseResult.headers.keys:
    filenames.incl(k)

  for filename in filenames:
    let missing = parseResult.allMissing.getOrDefault(filename)
    var fileDeps: Table[string, HashSet[string]]

    for otherFile in filenames:
      if otherFile == filename:
        continue

      let provides = parseResult.allProvides.getOrDefault(otherFile)

      # Direct matches
      var found: HashSet[string]
      for m in missing:
        if m in provides:
          found.incl(m)

      # Note: Python version handles (value, enum) tuples for enum values
      # In Nim models, enums are tracked differently - this may need adjustment
      # based on how the parser populates provides/missing sets

      if found.len > 0:
        fileDeps[otherFile] = found

    result[filename] = fileDeps


proc resolveShortName(shortName: string, allNames: seq[string]): string =
  ## Resolve a short type name to its fully qualified form.
  ##
  ## Args:
  ##   shortName: Short type name like "Point"
  ##   allNames: All fully qualified names like "ldtk::Point"
  ##
  ## Returns:
  ##   Fully qualified name if found, otherwise empty string.
  if shortName in allNames:
    return shortName

  # Try to find a match ending with ::shortName
  let suffix = "::" & shortName
  for name in allNames:
    if name.endsWith(suffix):
      return name

  return ""


proc getFieldDeps(fields: seq[FieldDecl]): seq[string] =
  ## Extract type dependencies from struct/class fields.
  for field in fields:
    let deps = getTemplateDependencies(field.typeName)
    for dep in deps:
      result.add(dep)


proc findTypeDependencies(self: DependencyAnalyzer, typeName: string, parseResult: ParseResult,
                          visited: var HashSet[string]): HashSet[string] =
  ## Find all types that a given type depends on.
  ##
  ## Args:
  ##   typeName: The type to find dependencies for.
  ##   parseResult: Parsed header data.
  ##   visited: Already visited types (for cycle detection).
  ##
  ## Returns:
  ##   Set of dependent type names.
  if typeName in visited:
    return

  visited.incl(typeName)

  # Collect all types from all headers
  var allClasses: seq[tuple[fullyQualified: string, underlyingDeps: seq[string]]]
  var allEnums: seq[string]

  for header in parseResult.headers.values:
    for s in header.structs:
      # Include both underlyingDeps and field type dependencies
      let fieldDeps = getFieldDeps(s.fields)
      allClasses.add((s.fullyQualified, s.underlyingDeps & s.baseTypes & fieldDeps))
    for c in header.classes:
      # Include base types and field type dependencies for classes
      let fieldDeps = getFieldDeps(c.fields)
      allClasses.add((c.fullyQualified, c.baseTypes & fieldDeps))
    for t in header.typedefs:
      # Include field dependencies if typedef has embedded struct
      var deps = t.underlyingDeps
      if t.structData.isSome:
        let fieldDeps = getFieldDeps(t.structData.get.fields)
        deps = deps & fieldDeps
      allClasses.add((t.fullyQualified, deps))
    for e in header.enums:
      allEnums.add(e.fullyQualified)

  var classesFully: seq[string]
  for c in allClasses:
    classesFully.add(c.fullyQualified)

  # Find the type and its dependencies
  for (fullyQualified, underlyingDeps) in allClasses:
    if fullyQualified != typeName:
      continue

    for dep in underlyingDeps:
      # Try to resolve short names to fully qualified names
      let resolvedDep = resolveShortName(dep, classesFully)
      let resolvedEnum = resolveShortName(dep, allEnums)

      if resolvedEnum.len > 0:
        result.incl(resolvedEnum)
      elif resolvedDep.len > 0:
        if resolvedDep notin result:
          let newDeps = self.findTypeDependencies(resolvedDep, parseResult, visited)
          result.incl(resolvedDep)
          for d in newDeps:
            result.incl(d)


proc computeSharedTypes(self: DependencyAnalyzer, parseResult: ParseResult,
                        relationships: Table[string, Table[string, HashSet[string]]]): HashSet[string] =
  ## Determine which types should be in the shared types file.
  var shared: HashSet[string]

  # Types needed by multiple files
  for deps in relationships.values:
    for types in deps.values:
      for t in types:
        shared.incl(t)

  # Add dependencies of shared types
  var allShared = shared
  for typeName in shared:
    var visited: HashSet[string]
    let deps = self.findTypeDependencies(typeName, parseResult, visited)
    for d in deps:
      allShared.incl(d)

  # Auto-promote types that appear in multiple files to avoid duplicate definitions
  var typeFiles: Table[string, int]  # fullyQualified -> file count
  for _, header in parseResult.headers:
    for s in header.structs:
      typeFiles.mgetOrPut(s.fullyQualified, 0).inc
    for e in header.enums:
      typeFiles.mgetOrPut(e.fullyQualified, 0).inc
    for t in header.typedefs:
      typeFiles.mgetOrPut(t.fullyQualified, 0).inc
  for typeName, count in typeFiles:
    if count > 1:
      allShared.incl(typeName)

  # Add forced shared types from config
  for t in self.config.forceSharedTypes:
    allShared.incl(t)

  result = allShared


proc getNewName(self: DependencyAnalyzer, fullName: string, existingNames: seq[string],
                rootNamespace: Option[string]): string =
  ## Generate a new unique name for a type.
  ##
  ## Args:
  ##   fullName: Fully qualified C++ name.
  ##   existingNames: Names already in use.
  ##   rootNamespace: Root namespace to strip.
  ##
  ## Returns:
  ##   New unique name.
  let parts = fullName.split("::")
  if rootNamespace.isSome and parts.len > 0 and parts[0] == rootNamespace.get:
    return parts[1..^1].join("_")
  return flattenNamespace(fullName)


proc computeRenames(self: DependencyAnalyzer, parseResult: ParseResult,
                    sharedTypes: HashSet[string]): Table[string, string] =
  ## Generate new names for types that would otherwise conflict.
  # Find types in shared that have the same short name
  var shortNames: Table[string, seq[string]]

  for typeName in sharedTypes:
    let short = typeName.split("::")[^1]
    if short notin shortNames:
      shortNames[short] = @[]
    shortNames[short].add(typeName)

  # Rename types with conflicting short names
  let rootNamespace = self.config.rootNamespace

  for short, fullNames in shortNames:
    if fullNames.len > 1:
      for full in fullNames:
        var existingNames: seq[string]
        for v in result.values:
          existingNames.add(v)
        let newName = self.getNewName(full, existingNames, rootNamespace)
        result[full] = newName


proc computeImports(self: DependencyAnalyzer, parseResult: ParseResult,
                    relationships: Table[string, Table[string, HashSet[string]]]): Table[string, HashSet[string]] =
  ## Build the import graph for generated files.
  for filename, deps in relationships:
    var fileImports: HashSet[string]
    for depFile in deps.keys:
      # Convert path to module name (replace dashes for valid Nim identifiers)
      let moduleName = extractFilename(depFile).changeFileExt("").replace("-", "_")
      fileImports.incl(moduleName)
    result[filename] = fileImports


proc computeBaseClasses(self: DependencyAnalyzer, parseResult: ParseResult): HashSet[string] =
  ## Find all classes that are used as base classes.
  ## These need `of RootObj` pragma for inheritance to work.
  for header in parseResult.headers.values:
    for cls in header.classes:
      for baseType in cls.baseTypes:
        # Clean the base type name (remove const, &, etc.)
        var baseName = baseType.strip()
        if baseName.startsWith("const "):
          baseName = baseName[6..^1].strip()
        # Remove template params for matching
        let bracketIdx = baseName.find('<')
        if bracketIdx >= 0:
          baseName = baseName[0..<bracketIdx]
        result.incl(baseName)

    for struct in header.structs:
      for baseType in struct.baseTypes:
        var baseName = baseType.strip()
        if baseName.startsWith("const "):
          baseName = baseName[6..^1].strip()
        let bracketIdx = baseName.find('<')
        if bracketIdx >= 0:
          baseName = baseName[0..<bracketIdx]
        result.incl(baseName)

  # Also include types from config.inheritableTypes
  for t in self.config.inheritableTypes:
    result.incl(t)


proc analyze*(self: DependencyAnalyzer, parseResult: ParseResult): AnalysisResult =
  ## Perform full dependency analysis.
  ##
  ## Args:
  ##   parseResult: The result from parsing headers.
  ##
  ## Returns:
  ##   AnalysisResult with dependency information.
  result = initAnalysisResult()

  # Compute file relationships
  result.fileRelationships = self.computeRelationships(parseResult)

  # Identify shared types
  result.sharedTypes = self.computeSharedTypes(parseResult, result.fileRelationships)

  # Find base classes that need of RootObj
  result.baseClasses = self.computeBaseClasses(parseResult)

  # Base classes should be shared types (they're used by multiple derived classes)
  for baseClass in result.baseClasses:
    result.sharedTypes.incl(baseClass)

  # Compute type renames
  result.typeRenames = self.computeRenames(parseResult, result.sharedTypes)

  # Build import graph
  result.importGraph = self.computeImports(parseResult, result.fileRelationships)


# Legacy format support functions

proc legacyRelationships*(data: seq[tuple[filename: string, idx: int, kind: string, extra: string]],
                          provides: Table[string, HashSet[string]],
                          missing: Table[string, HashSet[string]]): Table[string, Table[string, HashSet[string]]] =
  ## Compute file relationships from legacy format.
  var filenames: HashSet[string]
  for item in data:
    filenames.incl(item.filename)

  for filename in filenames:
    let fileMissing = missing.getOrDefault(filename)
    var fileDeps: Table[string, HashSet[string]]

    for other in filenames:
      if other == filename:
        continue

      let fileProvides = provides.getOrDefault(other)
      var found: HashSet[string]
      for m in fileMissing:
        if m in fileProvides:
          found.incl(m)

      if found.len > 0:
        fileDeps[other] = found

    result[filename] = fileDeps


proc findDependenciesLegacy*(obj: string, classesInfo: seq[tuple[fullyQualified: string, underlyingDeps: seq[string]]],
                             enumsFully: seq[string], classesFully: seq[string],
                             visited: var HashSet[string]): HashSet[string] =
  ## Find all dependencies of a type (simplified interface).
  ##
  ## Args:
  ##   obj: Type name to find dependencies for.
  ##   classesInfo: Sequence of (fullyQualified, underlyingDeps) for classes/structs/typedefs.
  ##   enumsFully: Sequence of fully qualified enum names.
  ##   classesFully: Sequence of fully qualified class names.
  ##   visited: Recursion tracking set.
  ##
  ## Returns:
  ##   Set of dependency type names.
  if obj in visited:
    return

  visited.incl(obj)

  for (fullyQualified, underlyingDeps) in classesInfo:
    if fullyQualified != obj:
      continue

    for dep in underlyingDeps:
      if dep in enumsFully:
        result.incl(dep)
      elif dep in classesFully:
        if dep notin result:
          let newDeps = findDependenciesLegacy(dep, classesInfo, enumsFully, classesFully, visited)
          result.incl(dep)
          for d in newDeps:
            result.incl(d)


proc collectSharedObjects*(relationships: Table[string, Table[string, HashSet[string]]],
                           forceShared: seq[string] = @[]): HashSet[string] =
  ## Collect objects that need to be moved to shared types file.
  ##
  ## Args:
  ##   relationships: File relationship mapping.
  ##   forceShared: Types to force into shared file.
  ##
  ## Returns:
  ##   Set of type names to move to shared file.
  for deps in relationships.values:
    for typeSet in deps.values:
      for t in typeSet:
        result.incl(t)

  for t in forceShared:
    result.incl(t)


proc topoSortTypes*(entries: seq[SharedTypeEntry]): seq[SharedTypeEntry] =
  ## Topologically sort types so dependencies come first.
  ## Generic types come before their instantiations.
  ## Base classes come before derived classes.
  var nameToEntry: Table[string, SharedTypeEntry]
  var inDegree: Table[string, int]
  var graph: Table[string, seq[string]]  # name -> types that depend on it

  # Build lookup and initialize
  for entry in entries:
    nameToEntry[entry.name] = entry
    inDegree[entry.name] = 0
    graph[entry.name] = @[]

  # Build dependency graph
  for entry in entries:
    for dep in entry.deps:
      # Check if dep matches any entry (handle namespace stripping)
      for name in nameToEntry.keys:
        if name == dep or name.endsWith("::" & dep) or dep.endsWith("::" & name.split("::")[^1]):
          if name != entry.name:
            graph[name].add(entry.name)
            inDegree[entry.name] = inDegree.getOrDefault(entry.name) + 1

  # Kahn's algorithm with priority:
  # 1. Generic types (isGeneric) first
  # 2. Base classes second
  # 3. Then by dependency order
  var ready: seq[string]
  for name, degree in inDegree:
    if degree == 0:
      ready.add(name)

  # Sort ready queue: generics first, then base classes
  ready.sort do (a, b: string) -> int:
    let ea = nameToEntry[a]
    let eb = nameToEntry[b]
    if ea.isGeneric != eb.isGeneric:
      return if ea.isGeneric: -1 else: 1
    if ea.isBaseClass != eb.isBaseClass:
      return if ea.isBaseClass: -1 else: 1
    return cmp(a, b)

  var sortedResult: seq[SharedTypeEntry]
  while ready.len > 0:
    let name = ready[0]
    ready.delete(0)
    sortedResult.add(nameToEntry[name])

    for dependent in graph.getOrDefault(name):
      inDegree[dependent] = inDegree[dependent] - 1
      if inDegree[dependent] == 0:
        # Insert maintaining sort order
        var inserted = false
        for i, r in ready:
          let er = nameToEntry[r]
          let ed = nameToEntry[dependent]
          if (ed.isGeneric and not er.isGeneric) or
             (ed.isBaseClass and not er.isBaseClass and not er.isGeneric):
            ready.insert(dependent, i)
            inserted = true
            break
        if not inserted:
          ready.add(dependent)

  # Add any remaining (cycles) at the end
  for entry in entries:
    if entry.name notin sortedResult.mapIt(it.name):
      sortedResult.add(entry)

  return sortedResult
